pkg/controller/clusterpool/collections.go

```diff
@@ -584,6 +599,252 @@ func (cds *cdCollection) Delete(c client.Client, cdName string) error {
 	return nil
 }
 
+type cdcCollection struct {
+	// Unclaimed by any cluster pool CD any not broken
+	unassigned []*hivev1.ClusterDeploymentCustomization
+	// Listed in the cluster pool inventory but not found
+	missing []string
+	// Used by some cluster deployment
+	reserved []*hivev1.ClusterDeploymentCustomization
+	// Last Cluster Deployment failed on provision
+	cloud []*hivev1.ClusterDeploymentCustomization
+	// Failed to apply patches for this cluster pool
+	syntax []*hivev1.ClusterDeploymentCustomization
+	// All CDCs in this pool
+	byCDCName map[string]*hivev1.ClusterDeploymentCustomization
+}
+
+// getAllCustomizationsForPool is the constructor for a cdcCollection for all of the
+// ClusterDeploymentCustomizations that are related to specified pool.
+func getAllCustomizationsForPool(c client.Client, pool *hivev1.ClusterPool, logger log.FieldLogger) (*cdcCollection, error) {
+	if pool.Spec.Inventory == nil {
+		return nil, nil
+	}
+	cdcList := &hivev1.ClusterDeploymentCustomizationList{}
+	if err := c.List(
+		context.Background(), cdcList,
+		client.MatchingFields{claimClusterPoolIndex: pool.Name},
+		client.InNamespace(pool.Namespace)); err != nil {
+		logger.WithError(err).Error("error listing ClusterDeploymentCustomizations")
+		return nil, err
+	}
+
+	cdcCol := cdcCollection{
+		unassigned: make([]*hivev1.ClusterDeploymentCustomization, 0),
+		missing:    make([]string, 0),
+		reserved:   make([]*hivev1.ClusterDeploymentCustomization, 0),
+		cloud:      make([]*hivev1.ClusterDeploymentCustomization, 0),
+		syntax:     make([]*hivev1.ClusterDeploymentCustomization, 0),
+		byCDCName:  make(map[string]*hivev1.ClusterDeploymentCustomization),
+	}
+
+	for _, item := range pool.Spec.Inventory {
+		missing := true
+		for i, cdc := range cdcList.Items {
```

Comment:
I thought I mentioned this already, but maybe it was for something else.

We can make this chunk O(N) rather than O(N²) by first looping through just the CDCs and putting them in a (temporary) `map[cdcName]cdc`. Then the second loop spins through the Inventory and looks up the CDC from that temporary map, adding it to the permanent `byCDCName` map and shunting it into the appropriate state-specific bucket (or adding the name to `missing` if it's not in the temporary map).

This isn't a functional, but it'll make the code a bit more efficient. And (IMO) a bit easier to read, particularly wrt the `missing` logic.